File: src/main/java/nl/insomnia247/nailbiter/eldrinkopubbot/eldrinko/action/ElDrinkoActionInflator.java
package nl.insomnia247.nailbiter.eldrinkopubbot.eldrinko.action;
import org.apache.commons.collections4.IterableUtils;
import org.apache.commons.collections4.Predicate;
import java.util.Arrays;
import java.util.function.Function;
import nl.insomnia247.nailbiter.eldrinkopubbot.util.MiscUtils;
import org.bson.Document;
import com.mongodb.client.MongoCollection;
import nl.insomnia247.nailbiter.eldrinkopubbot.eldrinko.ElDrinkoInputMessage;
import java.util.Date;
import nl.insomnia247.nailbiter.eldrinkopubbot.telegram.TelegramTextInputMessage;
import nl.insomnia247.nailbiter.eldrinkopubbot.telegram.TelegramImageOutputMessage;
import org.apache.commons.collections4.ListUtils;
import java.util.TimeZone;
import nl.insomnia247.nailbiter.eldrinkopubbot.telegram.TelegramKeyboard;
import org.apache.commons.lang3.tuple.ImmutablePair;
import java.util.Map;
import java.util.List;
import nl.insomnia247.nailbiter.eldrinkopubbot.util.Tsv;
import nl.insomnia247.nailbiter.eldrinkopubbot.util.JSONTools;
import java.util.ArrayList;
import java.net.URL;
import java.util.HashMap;
import nl.insomnia247.nailbiter.eldrinkopubbot.util.TemplateEngine;
import nl.insomnia247.nailbiter.eldrinkopubbot.util.MiscUtils;
import nl.insomnia247.nailbiter.eldrinkopubbot.telegram.TelegramInputMessage;
import org.json.JSONObject;
import org.json.JSONArray;
import nl.insomnia247.nailbiter.eldrinkopubbot.model.OutputMessage;
import nl.insomnia247.nailbiter.eldrinkopubbot.util.MiscUtils;
import nl.insomnia247.nailbiter.eldrinkopubbot.telegram.UserData;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import java.util.function.Consumer;
import nl.insomnia247.nailbiter.eldrinkopubbot.util.PersistentStorage;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import nl.insomnia247.nailbiter.eldrinkopubbot.telegram.TelegramTextOutputMessage;
import nl.insomnia247.nailbiter.eldrinkopubbot.model.OutputArrayMessage;
import nl.insomnia247.nailbiter.eldrinkopubbot.telegram.TelegramKeyboardAnswer;
import nl.insomnia247.nailbiter.eldrinkopubbot.util.SecureString;

/**
 * @author Alex Leontiev
 */
public class ElDrinkoActionInflator implements Function<Object,Function<ElDrinkoInputMessage,ImmutablePair<OutputMessage,JSONObject>>> {
    public static String[] BOTTLE_TYPES = new String[] {"0,5","1,0","1,5","2,0","3,0"};
    private static Logger _Log = LogManager.getLogger(ElDrinkoActionInflator.class);
    private Consumer<Document> _orderInserter;
    private static final JSONObject _TRANSITIONS 
        = new JSONObject(MiscUtils.GetResource("transitions",".json"));
    private final PersistentStorage _masterPersistentStorage;
    private JSONObject _config;
    private static final DateFormat _ORDER_REPORT_FORMATTER = new SimpleDateFormat("dd.MM.yy HH:mm");
    static {
        _ORDER_REPORT_FORMATTER.setTimeZone(TimeZone.getTimeZone("Ukraine/Kiev"));
    }
    private final Consumer<ImmutablePair<String,String>>  _sendOrderCallback;
    public ElDrinkoActionInflator(Consumer<ImmutablePair<String,String>> sendOrderCallback, PersistentStorage masterPersistentStorage
            , Consumer<Document> orderInserter) {
        _masterPersistentStorage = masterPersistentStorage;
        _sendOrderCallback = sendOrderCallback;
        _orderInserter = orderInserter;
        if(_orderInserter == null) {
            _orderInserter = new Consumer<Document> () {
                @Override
                public void accept(Document d) {}
            };
        }
    }
    @Override
    public Function<ElDrinkoInputMessage,ImmutablePair<OutputMessage,JSONObject>> apply(Object o) {
        return new Function<ElDrinkoInputMessage, ImmutablePair<OutputMessage,JSONObject>>() {
            @Override
            public ImmutablePair<OutputMessage,JSONObject> apply(ElDrinkoInputMessage im) {
                _Log.info(SecureString.format("here with %s,%s",o,im));
                if( (((JSONObject)o).getString("correspondence")).equals("9c851972cb7438c5") || (((JSONObject)o).getString("correspondence")).equals("07defdb4543782cb")) {
                    _Log.info(SecureString.format("%s",o));
                    if ((((JSONObject)o).optString("src_state")).equals("choose_product_to_make_order")) {
                        TelegramKeyboardAnswer tka = (TelegramKeyboardAnswer) im.left;
                        int i = Integer.parseInt(tka.getMsg());
                        Tsv tsv = im.beerlist;
                        JSONObject order = im.right.optJSONObject("order");
                        if(order==null) {
                            order = new JSONObject();
                            order.put("cart",new JSONArray());
                        }
                        JSONObject obj = new JSONObject();
                        String name = tsv.getColumn("name").get(i);
                        //FIXME: remove `amount` from here
                        obj.put("name",name).put("bottles",new JSONObject()).put("amount",0.0f);
                        order.getJSONArray("cart").put(obj);
                        im.right.put("order",order);
                    } else if( ((JSONObject)o).optString("type").equals("validButton") ) {
                        _Log.info(SecureString.format("%s",im.left.getMsg()));
                        _Log.info(SecureString.format("%s",im.right));
                        JSONArray cart = im.right.getJSONObject("order").getJSONArray("cart");
                        JSONObject lastItem = cart.getJSONObject(cart.length()-1);
                        JSONObject bottles = lastItem.getJSONObject("bottles");
                        int _i = Integer.parseInt(im.left.getMsg());
                        _Log.info(SecureString.format("_i: %d",_i));
                        int idx = _i/4;
                        _Log.info(SecureString.format("idx: %d",idx));
                        boolean shouldAdd = _i%4==1;
                        _Log.info(SecureString.format("shouldAdd: %s",shouldAdd));
                        String bottleType = BOTTLE_TYPES[idx];

                        if( !bottles.has(bottleType) ) {
                            bottles.put(bottleType,0);
                        }
                        bottles.put(bottleType,bottles.getInt(bottleType) + (shouldAdd?1:-1));
                        bottles.put(bottleType,Math.max(bottles.getInt(bottleType),0));
                        float amount = 0.0f;
                        for(String s: BOTTLE_TYPES) {
                            try {
                                amount += MiscUtils.ParseFloat(s) * bottles.optInt(s,0);
                            } catch (MiscUtils.ParseFloatException e) {
                                _Log.error(e);
                            }
                        }
                        lastItem.put("amount",amount);
                    }
                } else if( ((JSONObject)o).getString("correspondence").equals("5e11c9696e9b38f0") ) {
                    if(((JSONObject)o).optString("src_state").equals("delete")) {
                        TelegramKeyboardAnswer tka = (TelegramKeyboardAnswer) im.left;
                        int i = Integer.parseInt(tka.getMsg());
                        JSONObject order = im.right.optJSONObject("order");
                        JSONArray cart = order.getJSONArray("cart");
                        JSONObject removed = cart.getJSONObject(i);
                        cart.remove(i);
                    }
                } else if( (((JSONObject)o).getString("correspondence").equals("72e97b89bcab08c4") && ((JSONObject)o).getString("src_state").equals("choose_address")) ||  ((JSONObject)o).getString("correspondence").equals("774ed3e0f5ef17cf")) {
                    im.right.put("address",im.left.getMsg());
                } else if( ((JSONObject)o).getString("correspondence").equals("8e0edde4a3199d0c") ) {
                    im.right.put("phone_number",im.left.getMsg());
                } else if( ((JSONObject)o).getString("correspondence").equals("fa702a44b70ddcae") ) {
                    if(((JSONObject)o).optString("src_state").equals("edit_address")) {
                        im.right.put("address",im.left.getMsg());
                    } else if(((JSONObject)o).optString("src_state").equals("choose_payment")) {
                        TelegramKeyboardAnswer tka = (TelegramKeyboardAnswer) im.left;
                        int i = Integer.parseInt(tka.getMsg());
                        String paymentMethods 
                            = MiscUtils.ProcessTemplate("4ea9a63509e8ed5826a37f8a",null,im.beerlist);
                        im.right.put("payment", paymentMethods.split("\n")[i]);
                    } else if(((JSONObject)o).optString("src_state").equals("edit_phone_number")) {
                        im.right.put("phone_number",im.left.getMsg());
                    }
                } else if( ((JSONObject)o).getString("correspondence").equals("48c6907046b03db8") ) {
                    JSONObject order = _GetOrder(im.right);
                    order.put("uid",im.userData.getUserName());
                    order.put("count",_IncrementOrderCount(_masterPersistentStorage));
                    Date d = new Date();
                    order.put("timestamp", _ORDER_REPORT_FORMATTER.format(d));
                    order.put("_timestamp", d);
                    Map<String,Object> map = _OrderObjectToJinjaContext(order,im.beerlist);
                    _orderInserter.accept(new Document(map));
                    _sendOrderCallback.accept(new ImmutablePair<String,String>(
                            MiscUtils.ProcessTemplate("3804e512b18b339fe8786dbd",map,im.beerlist)
                                ,"salesmanChatIds"));
                    _sendOrderCallback.accept(new ImmutablePair<String,String>(
                            MiscUtils.ProcessTemplate("3804e512b18b339fe8786dbd",map,im.beerlist)
                            , im.userData.getChatId().toString()
                                ));
                    im.right.put("order","");
                } else if( ((JSONObject)o).getString("correspondence").equals("75f676187e00dd85") && ((JSONObject)o).optString("src_state").equals("null") ) {
                    im.right.put("order","");
                }

                _Log.info(SecureString.format("before _GetOrder(%s)",im.right));
                JSONObject ppo = _GetOrder(im.right);
                _Log.info(SecureString.format("_OrderObjectToJinjaContext(%s,%s)",ppo,im.beerlist.toJsonString()));
                Map<String,Object> map = _OrderObjectToJinjaContext(ppo,im.beerlist);
                Object oo = null;
                _Log.info(SecureString.format("after _GetOrder(%s)",im.right));

                if( (((JSONObject)o).getString("correspondence")).equals("72aa7197071b6503") ) {
                    TelegramKeyboardAnswer tka = (TelegramKeyboardAnswer) im.left;
                    Tsv tsv = im.beerlist;
                    String imgUrl 
                        = tsv.getColumn("image link").get(Integer.parseInt(tka.getMsg()));
                    map.put("i",Integer.parseInt(tka.getMsg()));
                    oo = MiscUtils.SafeUrl(imgUrl);
                } else if((((JSONObject)o).getString("correspondence")).equals("9c851972cb7438c5") || (((JSONObject)o).getString("correspondence")).equals("07defdb4543782cb") ) {
                    Map<String,Object> beerVolumes = new HashMap<>();
                    Tsv tsv = im.beerlist;

                    float totalVolume = 0.0f;

                    JSONArray cart = im.right.getJSONObject("order").getJSONArray("cart");
                    JSONObject bottles = cart.getJSONObject(cart.length()-1).getJSONObject("bottles");
                    String beerName = cart.getJSONObject(cart.length()-1).getString("name");
                    for(String s:BOTTLE_TYPES) {
                        float volume = 0.0f;
                        try {
                            volume = MiscUtils.ParseFloat(s);
                        } catch (MiscUtils.ParseFloatException e) {
                            _Log.error(e);
                        }
                        totalVolume += volume*bottles.optInt(s);
                    }
                    int i = IterableUtils.indexOf(tsv.getColumn("name"),new Predicate<String>(){
                        @Override
                        public boolean evaluate(String n) {
                            return n.equals(beerName);
                        }
                    });
                    float totalPrice = 0.0f;
                    try {
                        totalPrice = totalVolume*MiscUtils.ParseFloat(tsv.getColumn("price (UAH/L)").get(i));
                    } catch (MiscUtils.ParseFloatException e) {
                        _Log.error(e);
                    }
                    map.put("totalVolume",totalVolume);
                    map.put("totalPrice",totalPrice);
                    _Log.info(SecureString.format("map: %s",map));
                }

                im.right.put("username",im.userData.getUserName());
                im.right.put("firstName",im.userData.getFirstName());
                im.right.put("lastName",im.userData.getLastName());
                _Log.info(SecureString.format("before _InflateOutputMessage(%s,%s,%s,%s),%s",((JSONObject)o).getString("correspondence"),map,oo,im.beerlist.toJsonString(),im.right));
                OutputMessage om = _InflateOutputMessage(((JSONObject)o).getString("correspondence"),map,oo,im.beerlist);
                _Log.info(SecureString.format("om: %s",om));
                return new ImmutablePair<OutputMessage,JSONObject>(om,im.right);
            }
        };
    }
    private static OutputMessage _InflateOutputMessage(String code, Map<String,Object> env, Object obj, Tsv tsv) {
        Object m = _TRANSITIONS.getJSONObject("transitions").get(code);
        _Log.info(SecureString.format("m: %s",m));
        return _InflateOutputMessageFromJson(m,env,obj,tsv);
    }
    private static OutputMessage _InflateOutputMessageFromJson(Object m, Map<String,Object> env, Object obj,Tsv tsv) {
        assert m instanceof JSONObject || m instanceof JSONArray;
        _Log.info(m);
        if(m instanceof JSONArray) {
            List<OutputMessage> msgs = new ArrayList<>();

            _Log.info(SecureString.format(" %s ",m));
            for(int i = 0; i < ((JSONArray)m).length();i++) {
                msgs.add(_InflateOutputMessageFromJson(((JSONArray)m).get(i),env,obj,tsv));
            }
            return new OutputArrayMessage(msgs.toArray(OutputMessage[]::new));
        } else {
            JSONObject msg = (JSONObject)m;
            String tag = msg.getString("tag");
            if(tag.equals("TelegramTextOutputMessage")) {
                return new TelegramTextOutputMessage(MiscUtils.ProcessTemplate(msg.getString("message"), env,tsv));
            } else if(tag.equals("TelegramKeyboard")) {
                return _InflateTelegramKeyboard(
                        env, msg.getString("message"), msg.getString("keyboard"),tsv,msg.optInt("columns",2));
            } else if(tag.equals("TelegramImageOutputMessage")) {
                return new TelegramImageOutputMessage(
                        MiscUtils.ProcessTemplate(msg.getString("message"), env,tsv)
                        , (URL)obj);
            } else {
                return null;
            }
        }
    }
    private static Map<String,Object> _OrderObjectToJinjaContext(JSONObject order, Tsv tsv) {
        Map<String, Object> context = new HashMap<String,Object>();
        context.put("BOTTLES",Arrays.asList(BOTTLE_TYPES));

        _Log.info(order);
        if(order==null) {
            return context;
        }
        List<List<String>> products = tsv.getRecords();
        Map<String,Object> orderMap = JSONTools.JSONObjectToMap(order);
        //FIXME: try to compute `sum` in templates
        float sum = 0;
        JSONArray cart = order.getJSONArray("cart");
        for(int i = 0; i < cart.length(); i++) {
            JSONObject obj = cart.getJSONObject(i);
            if(!obj.has("amount")) {
                continue;
            }
            float beerPrice = Float.parseFloat(products.stream()
                    .filter(r -> r.get(1).equals(obj.getString("name")))
                    .findAny()
                    .orElse(null)
                    .get(3));
            sum += beerPrice * obj.getDouble("amount");
        }
        orderMap.put("sum",sum);
        orderMap.put("delivery_fee",sum>=250 ? (double)0.0 : (double)20.0);
        _Log.info(orderMap.toString());
        context.put("order",orderMap);
        _Log.info(SecureString.format("context: %s",context));
        return context;
    }
    private static TelegramKeyboard _InflateTelegramKeyboard(Map<String,Object> env,String msgTemplateResName, String keysTemplateResName,Tsv tsv,int columns) {
        String keyboardKeys = MiscUtils.ProcessTemplate(keysTemplateResName,env,tsv);
        String keyboardMsg = MiscUtils.ProcessTemplate(msgTemplateResName,env,tsv);
        _Log.info(SecureString.format("keyboardMsg: %s",keyboardMsg));
        _Log.info(SecureString.format("keyboardKeys: %s",keyboardKeys));
        return new TelegramKeyboard(keyboardMsg,keyboardKeys.split("\n"),columns);
    }
    private static int _IncrementOrderCount(PersistentStorage masterPersistentStorage) {
        final String FN = "order_count";
        int res = 0;
        if( masterPersistentStorage.contains(FN) ) {
            res = Integer.parseInt(masterPersistentStorage.get(FN));
        }
        masterPersistentStorage.set(FN,Integer.toString(res+1));
        return res+1;
    }
    private static JSONObject _GetOrder(JSONObject obj) {
        JSONObject order = obj.optJSONObject("order");
        if(order!=null) {
            order = new JSONObject(order.toString());
            final String[] KEYS = new String[] {"phone_number","address","payment"};
            for(String k: KEYS) {
                if(obj.has(k)) {
                    order.put(k,obj.getString(k));
                }
            }
        }
        return order;
    }
}

File: nl/insomnia247/nailbiter/eldrinkopubbot/eldrinko/action/el_drinko_action_inflator.py
"""===============================================================================
        FILE: /Users/nailbiter/Documents/forgithub/ElDrinkoPubBot/nl/insomnia247/nailbiter/eldrinkopubbot/eldrinko/action/el_drinko_action_inflator.py
       USAGE: (not intended to be directly executed)
 DESCRIPTION:
     OPTIONS: ---
REQUIREMENTS: ---
        BUGS: ---
       NOTES: ---
      AUTHOR: Alex Leontiev (alozz1991@gmail.com)
ORGANIZATION:
     VERSION: ---
     CREATED: 2021-02-07T14:32:53.623252
    REVISION: ---
==============================================================================="""
from nl.insomnia247.nailbiter.eldrinkopubbot.eldrinko import ElDrinkoInputMessage
from nl.insomnia247.nailbiter.eldrinkopubbot.util import add_logger, ukrainian_floats
from nl.insomnia247.nailbiter.eldrinkopubbot.util.el_drinko_jinja_environment import ElDrinkoJinjaEnvironment, BOTTLE_TYPES
from nl.insomnia247.nailbiter.eldrinkopubbot.telegram import TelegramKeyboard, TelegramTextOutputMessage, TelegramImageOutputMessage, TelegramArrayOutputMessage
from jinja2.loaders import FileSystemLoader
import json
from datetime import datetime, date
import pytz
import logging
import math


def _increment_order_count(persistent_storage):
    res = int(persistent_storage.get("order_count", 0))
    res += 1
    persistent_storage.set("order_count", res)
    return res


class _DateTimeFormatter:
    def __init__(self):
        self._tz = pytz.timezone("Europe/Kiev")

    def __call__(self, d):
        return d.astimezone(self._tz).strftime("%Y.%m.%d %H:%M")


class ElDrinkoActionInflator:
    def __init__(self, send_message_callback, persistent_storage, insert_order_callback, template_folder):
        self._send_message_callback = send_message_callback
        self._persistent_storage = persistent_storage
        self._insert_order_callback = insert_order_callback
        self._sendOrderCallback = send_message_callback
        self._jinja_env = ElDrinkoJinjaEnvironment(
            loader=FileSystemLoader(template_folder))
        self._date_time_formatter = _DateTimeFormatter()
        self._transitions = None
        self._logger = logging.getLogger(self.__class__.__name__)
        self._transitions = {}
        with open(f"{template_folder}/transitions.json") as f:
            self._transitions["transitions"] = json.load(f)
        with open(f"{template_folder}/correspondence.json") as f:
            self._transitions["correspondence"] = json.load(f)

    @property
    def transitions(self):
        return self._transitions

    def _call(self, o, im, src_state, dst_state):
        """return (outmessage, user_data_update)"""
        if o["correspondence"] in ["9c851972cb7438c5", "07defdb4543782cb"]:
            if o.get("src_state", None) == "choose_product_to_make_order":
                tka = im.input_message
                i = int(tka.message)
                tsv = im.beerlist
                order = im.data.order
                obj = {}
                name = list(tsv["name"])[i]
                obj = {
                    **obj,
                    "name": name,
                    "bottles": {},
                }
                cart = order["cart"]
                cart.append(obj)
            elif o["type"] == "validButton":
                cart = im.data.order["cart"]
                lastItem = cart[-1]
                bottles = lastItem["bottles"]
                _i = int(im.input_message.message)
                idx = int(_i/4)
                shouldAdd = _i % 4 == 1
                bottleType = BOTTLE_TYPES[idx]
                if bottleType not in bottles:
                    bottles[bottleType] = 0
                bottles[bottleType] = bottles[bottleType] + \
                    (1 if shouldAdd else -1)
                bottles[bottleType] = max(bottles[bottleType], 0)
        elif o["correspondence"] == "5e11c9696e9b38f0":
            self._logger.info("here")
            if o.get("src_state", None) == "delete":
                self._logger.info("here")
                tka = im.input_message
                i = int(tka.message)
                order = im.data.order
                cart = order["cart"]
#                removed = cart[i]
#                cart = [x for i_, x in enumerate(cart) if i_ != i]
                del cart[i]
        elif (o["correspondence"] == "72e97b89bcab08c4" and o.get("src_state", None) == "choose_address") or o["correspondence"] == "774ed3e0f5ef17cf":
            im.data["address"] = im.input_message.message
        elif o["correspondence"] == "8e0edde4a3199d0c":
            im.data["phone_number"] = im.input_message.message
        elif o["correspondence"] == "fa702a44b70ddcae":
            if o.get("src_state", None) == "edit_address":
                im.data["address"] = im.input_message.message
            elif o.get("src_state", None) == "choose_payment":
                tka = im.input_message
                i = int(tka.message)
                paymentMethods = self._jinja_env.get_template(
                    "4ea9a63509e8ed5826a37f8a.txt").render().split("\n")
                im.data["payment"] = paymentMethods[i]
            elif o.get("src_state", None) == "edit_phone_number":
                im.data["phone_number"] = im.input_message.message
        elif o["correspondence"] == "48c6907046b03db8":
            order = im.data.get_order_pretty()
            order["uid"] = im.user_data.username
            tsv = im.beerlist
            order["count"] = _increment_order_count(self._persistent_storage)
            d = datetime.now()
            order["timestamp"] = self._date_time_formatter(d)
            order["_timestamp"] = d
            map_ = self._order_object_to_jinja_context(
                order, im.beerlist)
            self._insert_order_callback(map_)
            for tgt in ["salesmanChatIds", str(im.user_data)]:
                self._sendOrderCallback(
                    self._jinja_env.process_template(
                        "made_order_notification", map_, tsv),
                    tgt
                )
            im.data.reset_order()
        elif o["correspondence"] == "75f676187e00dd85" and o.get("src_state", None) is None:
            im.data.reset_order()

        ppo = im.data.get_order_pretty()
        map_ = self._order_object_to_jinja_context(
            ppo, im.beerlist)
        oo = None

        if o["correspondence"] == "72aa7197071b6503":
            tka = im.input_message
            tsv = im.beerlist
            imgUrl = list(tsv["image link"])[int(tka.message)]
            map_["i"] = int(tka.message)
            oo = imgUrl
        elif o["correspondence"] == "02503b04d94259c5":
            r = im.beerlist.query(f"name=='{im.input_message.button_title}'").to_dict(
                orient="records")[0]
            oo = r["image link"]
            map_["description"] = r["description"]
        elif o["correspondence"] == "18ca55e51d11ba24":
            map_["category"] = im.input_message.button_title
        elif o["correspondence"] == "12f00bba97cabd0d":
            cart = im.data["cart"]
            if src_state == "choose_snack_amount":
                if o["type"] == "validButton":
                    button_idx = int(im.input_message.message)
                    last_order = cart[-1]
                    self._logger.info(last_order)
                    self._logger.info(im.beerlist)
                    goods_df = im.beerlist.query(
                        f"category==\"{last_order['category']}\"")
                    self._logger.info(goods_df)
                    self._logger.info(button_idx)
                    inc = 1 if button_idx % 4 == 1 else -1
                    goods_r = goods_df.to_dict(orient="records")[
                        math.floor(button_idx/4)]
                    last_order["goods"][goods_r["name"]] = last_order["goods"].get(
                        goods_r["name"], 0) + inc
                    last_order["goods"][goods_r["name"]] = max(
                        last_order["goods"][goods_r["name"]], 0)
            else:
                cart.append(
                    {"category": im.input_message.button_title, "goods": {}})

        # FIXME: we need this?
        im.data["username"] = im.user_data.username
        im.data["firstName"] = im.user_data.first_name
        im.data["lastName"] = im.user_data.last_name
        om = self._inflate_output_message(
            o["correspondence"], {**map_, "data": im.data}, oo, im.beerlist)
        return (om, im.data)

    def __call__(self, obj, src_state, dst_state):
        return lambda eim: self._call(obj, eim, src_state, dst_state)

    def _inflate_output_message(self, code, env, obj, tsv):
        self._logger.info(code)
        m = self._transitions["transitions"][code]
        return self._inflate_output_message_from_json(m, env, obj, tsv)

    def _inflate_telegram_keyboard(self, env, msgTemplateResName, keysTemplateResName, tsv, columns):
        keyboardKeys = self._jinja_env.process_template(
            keysTemplateResName, env, tsv)
        keyboardKeys = keyboardKeys.strip().split("\n")
        self._logger.debug(f"keyboardKeys: {keyboardKeys}")
        keyboardKeys = [l for l in keyboardKeys if len(l.strip()) > 0]
        self._logger.debug(f"keyboardKeys: {keyboardKeys}")
        keyboardMsg = self._jinja_env.process_template(
            msgTemplateResName, env, tsv)
        return TelegramKeyboard(message=keyboardMsg, keyboard=keyboardKeys, columns=columns)

    def _inflate_output_message_from_json(self, m, env, obj, tsv):
        assert isinstance(m, dict) or isinstance(m, list)
        if isinstance(m, list):
            msgs = []
            for mm in m:
                if str(mm) in self._transitions["transitions"]:
                    mm = self._transitions["transitions"][str(mm)]
                msgs.append(self._inflate_output_message_from_json(
                    mm, env, obj, tsv))
            return TelegramArrayOutputMessage(messages=msgs)
        else:
            msg = m
            tag = msg["tag"]
            if tag == "TelegramTextOutputMessage":
                return TelegramTextOutputMessage(self._jinja_env.process_template(msg["message"], env, tsv))
            elif tag == "TelegramKeyboard":
                return self._inflate_telegram_keyboard(env, msg["message"], msg["keyboard"], tsv, msg.get("columns", 2))
            elif tag == "TelegramImageOutputMessage":
                return TelegramImageOutputMessage(message=self._jinja_env.process_template(msg["message"], env, tsv), url=obj)
            else:
                return None

    def _order_object_to_jinja_context(self, order, tsv):
        self._logger.info(f"order: {order}")
        context = {}
        if order is not None:
            context["order"] = order
        self._logger.info(f"context: {context}")
        return context
